TodoListWebApp/App_Start/Startup.Auth.cs

```diff
@@ -32,53 +56,87 @@ public void ConfigureAuth(IAppBuilder app)
                             // Remember that the base URL of the address used here must be provisioned in Azure AD beforehand.
-                            string appBaseUrl = context.Request.Scheme + "://" + context.Request.Host + context.Request.PathBase;                         
+                            string appBaseUrl = context.Request.Scheme + "://" + context.Request.Host + context.Request.PathBase + "/";
                             context.ProtocolMessage.RedirectUri = appBaseUrl;
                             context.ProtocolMessage.PostLogoutRedirectUri = appBaseUrl;
                             return Task.FromResult(0);
                         },
-                        // we use this notification for injecting our custom logic
                         SecurityTokenValidated = (context) =>
                         {
-                            // retriever caller data from the incoming principal
+                            // retrieve caller data from the incoming principal
                             string issuer = context.AuthenticationTicket.Identity.FindFirst("iss").Value;
-                            string UPN = context.AuthenticationTicket.Identity.FindFirst(ClaimTypes.Name).Value;
-                            string tenantID = context.AuthenticationTicket.Identity.FindFirst("http://schemas.microsoft.com/identity/claims/tenantid").Value;
+                            string Upn = context.AuthenticationTicket.Identity.FindFirst(ClaimTypes.Name).Value;
+                            string tenantId = context.AuthenticationTicket.Identity.FindFirst("http://schemas.microsoft.com/identity/claims/tenantid").Value;
 
                             if (
                                 // the caller comes from an admin-consented, recorded issuer
-                                (db.Tenants.FirstOrDefault(a => ((a.IssValue == issuer) && (a.AdminConsented))) == null)
-                                // the caller is recorded in the db of users who went through the individual onboardoing
-                                && (db.Users.FirstOrDefault(b =>((b.UPN == UPN) && (b.TenantID == tenantID))) == null)
+                                (this.db.Tenants.FirstOrDefault(a => ((a.IssValue == issuer) && (a.AdminConsented))) == null)
+                                // the caller is recorded in the db of users who went through the individual on-boarding
+                                && (this.db.Users.FirstOrDefault(b => ((b.UPN == Upn) && (b.TenantID == tenantId))) == null)
                                 )
                                 // the caller was neither from a trusted issuer or a registered user - throw to block the authentication flow
-                                throw new SecurityTokenValidationException();                            
+                                throw new SecurityTokenValidationException("Please use the Sign-up link to sign -up for the ToDo list application.");
```

Comment:
This is not really a SecurityTokenValidationException. We use this exception when we detect an invalid token, issues with: {expired, issuer, audience, signature }.
In this case the token is fine.
This is more of an authentication error.

Reply:
Changed to a UnauthorizedAccessException
thanks @brentschmaltz